pkg/controller/master/storagenetwork/storage_network.go

```
@@ -0,0 +1,851 @@
+		settingCopy.Annotations = make(map[string]string)
+	}
+
+	if ok, err := h.checkChange(settingCopy); !ok {
+		return nil, err
+	}
+
+	if earlyReturn, err := h.checkLonghornSetting(settingCopy); earlyReturn {
+		// err == nil, finish
+		// err != nil, reconcile
+		return nil, err
+	}
+
+	logrus.Infof("storage network change: %s", settingCopy.Value)
+
+	// if replica eq 0, skip
+	// save replica to annotation
+	// set replica to 0
+	if err := h.checkPodStatusAndStop(); err != nil {
+		logrus.Infof("Requeue to check pod status again")
+		if _, updateConditionErr := h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgStopPod); updateConditionErr != nil {
+			logrus.Errorf("update status error %v", updateConditionErr)
+			return nil, updateConditionErr
+		}
+		return nil, err
+	}
+
+	// check volume detach before put LH settings
+	if ok, err := h.checkLonghornVolumeDetached(); !ok {
+		if err != nil {
+			logrus.Errorf("check Longhorn volume error: %v", err)
+			return nil, err
+		}
+		logrus.Infof("still has attached volume")
+		if _, updateConditionErr := h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgWaitForVolumes); updateConditionErr != nil {
+			logrus.Errorf("update status error %v", updateConditionErr)
+			return nil, updateConditionErr
+		}
+		return nil, errors.New("Check Longhorn Volume status again, requeue")
+	}
+
+	logrus.Infof("all pods are stopped")
+	logrus.Infof("all volumes are detached")
+	logrus.Infof("update Longhorn settings")
+	// push LH setting
+	nadName := settingCopy.Annotations[NadStorageNetworkAnnotation]
+	if err := h.updateLonghornStorageNetwork(nadName); err != nil {
+		logrus.Errorf("Update Longhorn setting error %v", err)
+		return nil, err
+	}
+	if _, err := h.setConfiguredCondition(settingCopy, false, ReasonInProgress, MsgUpdateLonghornSetting); err != nil {
+		logrus.Errorf("update status error %v", err)
+		return nil, err
+	}
+
+	return nil, errors.New("Recheck Longhorn Setting and restart Pods.")
```

do not return the error at the end; just return the obj and nil to let the controller reconcile it.

---

This end line is not the config process's end.
So we still need to return an error to reconcile it directly.

---

fix